src/coreclr/System.Private.CoreLib/src/System/Runtime/ControlledExecution.CoreCLR.cs

```diff
@@ -0,0 +1,173 @@
+        [Obsolete(Obsoletions.ControlledExecutionRunMessage, DiagnosticId = Obsoletions.ControlledExecutionRunDiagId, UrlFormat = Obsoletions.SharedUrlFormat)]
+        public static void Run(Action action, CancellationToken cancellationToken)
+        {
+            ArgumentNullException.ThrowIfNull(action);
+            var execution = new Execution(action, cancellationToken);
+            cancellationToken.Register(execution.Abort, useSynchronizationContext: false);
+            execution.Run();
+        }
+
+        [LibraryImport(RuntimeHelpers.QCall, EntryPoint = "ThreadNative_Abort")]
+        private static partial void AbortThread(ThreadHandle thread);
+
+        [MethodImpl(MethodImplOptions.InternalCall)]
+        private static extern bool ResetAbortThread();
+
+        private sealed partial class Execution
+        {
+            // The state transition diagram (S means the Started flag and so on):
+            // N ⟶  S  ⟶ SF
+            // ↓     ↓
+            // AR   SAR ⟶ SFAR
+            // ↓     ↓       ↓
+            // A    SA  ⟶ SFA
+            private enum State : int
+            {
+                None = 0,
+                Started = 1,
+                Finished = 2,
+                AbortRequested = 4,
+                RunningAbort = 8
+            }
+
+            [ThreadStatic]
+            private static Execution? t_execution;
+
+            // Interpreted as a value of the State enumeration type
+            private int _state;
+            private readonly Action _action;
+            private readonly CancellationToken _cancellationToken;
+            private Thread? _thread;
+
+            public Execution(Action action, CancellationToken cancellationToken)
+            {
+                _action = action;
+                _cancellationToken = cancellationToken;
+            }
+
+            public void Run()
+            {
+                Debug.Assert((_state & (int)State.Started) == 0 && _thread == null);
+
+                // Recursive ControlledExecution.Run calls are not supported
+                if (t_execution != null)
+                    throw new InvalidOperationException(SR.InvalidOperation_NestedControlledExecutionRun);
+
+                _thread = Thread.CurrentThread;
+
+                try
+                {
+                    try
+                    {
+                        // As soon as the Started flag is set, this thread may be aborted asynchronously
+                        if (Interlocked.CompareExchange(ref _state, (int)State.Started, (int)State.None) == (int)State.None)
+                        {
+                            t_execution = this;
+                            _action();
+                        }
+                    }
+                    finally
+                    {
+                        if ((_state & (int)State.Started) != 0)
+                        {
+                            // Set the Finished flag to prevent a potential subsequent AbortThread call
```

Comment:
Why would there be a subsequent call?  No matter how many times the CancellationTokenSource associated with the CancellationToken has Cancel() called, the callback with the CancellationToken will only ever be invoked once.

Reply:
I meant that `Abort` may be running on another thread, which is going to call `AbortThread`.